File: backend/app/api/ripples.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user, get_optional_user
from app.models.user import User
from app.services.ripple_service import mark_delivery_consumed, touch_guest_session

router = APIRouter(prefix="/api/ripples", tags=["ripples"])


@router.post("/guest-session/touch")
async def touch_guest_ripple_session(
    request: Request,
    user: User | None = Depends(get_optional_user),
    db: AsyncSession = Depends(get_db),
):
    guest_session = await touch_guest_session(
        request.headers.get("x-ripple-guest-session"),
        db,
        claimed_user_id=user.id if user else None,
    )
    if guest_session is None:
        raise HTTPException(status_code=400, detail="Missing guest session header")

    return {
        "session_key": guest_session.session_key,
        "claimed_user_id": str(guest_session.claimed_user_id) if guest_session.claimed_user_id else None,
    }


@router.post("/deliveries/{delivery_id}/consume")
async def consume_ripple_delivery(
    delivery_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    push = await mark_delivery_consumed(delivery_id, user.id, db)
    if push is None:
        raise HTTPException(status_code=404, detail="Ripple delivery not found")
    return {"message": "Ripple delivery consumed", "delivery_id": str(push.id)}

File: apps/server/src/routes/ripples.ts
import { eq } from 'drizzle-orm';
import type { FastifyInstance } from 'fastify';
import type { AppDeps, Guards } from '../app.js';
import { ripplePushes } from '../db/schema.js';
import { AppError } from '../lib/errors.js';
import { guestSessionKey } from '../plugins/auth.js';
import { touchGuestSession } from '../services/ripple-service.js';

export function registerRippleRoutes(app: FastifyInstance, deps: AppDeps, guards: Guards): void {
  const { db } = deps;

  app.post('/api/ripples/guest-session/touch', async (request) => {
    const sessionKey = guestSessionKey(request);
    if (!sessionKey) throw AppError.badRequest('Missing or invalid guest session', 'invalid_session');
    await touchGuestSession(db, sessionKey);
    return { session_key: sessionKey, active: true };
  });

  async function updatePushStatus(
    deliveryId: string,
    userId: string,
    status: 'consumed' | 'dismissed',
  ) {
    const rows = await db
      .select()
      .from(ripplePushes)
      .where(eq(ripplePushes.id, deliveryId))
      .limit(1);
    const push = rows[0];
    if (!push || push.target_user_id !== userId) throw AppError.notFound('Delivery not found');
    await db
      .update(ripplePushes)
      .set({ status, ...(status === 'consumed' ? { consumed_at: new Date() } : {}) })
      .where(eq(ripplePushes.id, deliveryId));
    return { id: deliveryId, status };
  }

  app.post(
    '/api/ripples/deliveries/:deliveryId/consume',
    { preHandler: guards.requireAuth },
    async (request) => {
      const { deliveryId } = request.params as { deliveryId: string };
      return updatePushStatus(deliveryId, request.currentUser!.id, 'consumed');
    },
  );

  app.post(
    '/api/ripples/deliveries/:deliveryId/dismiss',
    { preHandler: guards.requireAuth },
    async (request) => {
      const { deliveryId } = request.params as { deliveryId: string };
      return updatePushStatus(deliveryId, request.currentUser!.id, 'dismissed');
    },
  );
}
